lib/Canonicalize/EmbedStatementComments.cpp

```diff
@@ -0,0 +1,163 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/IR/DebugInfoMetadata.h"
+#include "llvm/IR/IRBuilder.h"
+
+#include "revng/PTML/CommentPlacementHelper.h"
+#include "revng/Pipeline/LLVMContainer.h"
+#include "revng/Pipeline/Location.h"
+#include "revng/Pipeline/Pipe.h"
+#include "revng/Pipeline/RegisterPipe.h"
+#include "revng/Pipes/Kinds.h"
+#include "revng/Pipes/Ranks.h"
+#include "revng/Support/DecompilationHelpers.h"
+
+static std::optional<MetaAddress>
+tryExtractAddress(const llvm::Instruction &I) {
+  if (!I.getDebugLoc() || !I.getDebugLoc()->getScope())
+    return std::nullopt;
+
+  auto DebugLocation = I.getDebugLoc()->getScope()->getName().str();
+  auto Parsed = pipeline::locationFromString(revng::ranks::Instruction,
+                                             DebugLocation);
+  revng_assert(Parsed.has_value());
+
+  MetaAddress Extracted = Parsed->at(revng::ranks::Instruction);
+  revng_assert(Extracted.isValid());
+  return Extracted;
+}
+
+template<typename ResultSet>
+static ResultSet gatherNonStatementAddresses(const llvm::Instruction &I,
+                                             ResultSet &&Result = {}) {
+  if (std::optional MaybeAddress = tryExtractAddress(I))
+    Result.emplace(*MaybeAddress);
+
+  for (const llvm::Use &V : I.operands())
+    if (const llvm::Instruction *Cast = llvm::dyn_cast<llvm::Instruction>(V))
+      if (not isStatement(*Cast))
+        gatherNonStatementAddresses(*Cast, Result);
+
+  return Result;
+}
+
+template<>
+struct yield::StatementTraits<llvm::BasicBlock *> {
+
+  using StatementType = llvm::Instruction *;
+
+  static auto getStatements(llvm::BasicBlock *Node) {
+    return *Node | std::views::filter(isStatement)
+           | std::views::transform([](auto &&R) { return &R; });
+  }
+
+  static std::set<MetaAddress> getAddresses(StatementType Statement) {
+    return gatherNonStatementAddresses<std::set<MetaAddress>>(*Statement);
+  }
+};
+
+static std::string
+addressesToString(range_with_value_type<MetaAddress> auto const &Addresses) {
+  std::string Result = "";
+
+  if (not Addresses.empty()) {
+    for (const MetaAddress &Address : Addresses)
+      Result += Address.toString() + " + ";
+    Result.resize(Result.size() - 3);
+  }
+
+  return Result;
+}
+
+struct EmbedStatementComments {
+public:
+  static constexpr auto Name = "embed-statement-comments";
+
+private:
+  std::optional<llvm::FunctionCallee> LLVMIRComment = std::nullopt;
+  const llvm::FunctionCallee &llvmIRComment(llvm::Module &M) {
```

Comment:
Seems like you're trying not to define a function if it's not necessary.

For simplicity, this is not a practice we follow, let's create it in any case. Global DCE will purge it if unnecessary.

Reply:
> The reason this is in a separate helper [...]

When I mentioned "defining a function" I didn't mean the`llvmIRComment` function, but the `llvm::Function` you're creating (`comment`).

> As for the std::optional<llvm::FunctionCallee> thing

I'm saying: let's create it unconditionally before the first call to `llvmIRComment`.
There's no real benefit in not creating it. If we create it and then not use it, global dce will garbage collect it.